tools/stats/rpc.go

```diff
@@ -0,0 +1,169 @@
+package main
+
+import (
+	"context"
+	"log"
+	"net/http"
+	"time"
+
+	"github.com/multiformats/go-multiaddr-net"
+
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-lotus/api"
+	"github.com/filecoin-project/go-lotus/api/client"
+	"github.com/filecoin-project/go-lotus/chain"
+	"github.com/filecoin-project/go-lotus/chain/store"
+	"github.com/filecoin-project/go-lotus/chain/types"
+	"github.com/filecoin-project/go-lotus/lib/jsonrpc"
+	"github.com/filecoin-project/go-lotus/node/repo"
+)
+
+func getAPI(path string) (string, http.Header, error) {
+	r, err := repo.NewFS(path)
+	if err != nil {
+		return "", nil, err
+	}
+
+	ma, err := r.APIEndpoint()
+	if err != nil {
+		return "", nil, xerrors.Errorf("failed to get api endpoint: %w", err)
+	}
+	_, addr, err := manet.DialArgs(ma)
+	if err != nil {
+		return "", nil, err
+	}
+	var headers http.Header
+	token, err := r.APIToken()
+	if err != nil {
+		log.Printf("Couldn't load CLI token, capabilities may be limited: %w", err)
+	} else {
+		headers = http.Header{}
+		headers.Add("Authorization", "Bearer "+string(token))
+	}
+
+	return "ws://" + addr + "/rpc/v0", headers, nil
+}
+
+func WaitForSyncComplete(ctx context.Context, napi api.FullNode) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-time.After(30 * time.Second):
+			state, err := napi.SyncState(ctx)
+			if err != nil {
+				return err
+			}
+
+			log.Printf("Stage %s, Height %d", chain.SyncStageString(state.Stage), state.Height)
+
+			if state.Stage == api.StageSyncComplete {
+				return nil
+			}
+		}
+	}
+}
+
+func GetTips(ctx context.Context, api api.FullNode, lastHeight uint64) (<-chan *types.TipSet, error) {
+	chmain := make(chan *types.TipSet)
+
+	notif, err := api.ChainNotify(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	go func() {
+		defer close(chmain)
+
+		ping := time.Tick(30 * time.Second)
+
+		for {
+			select {
+			case changes := <-notif:
+				for _, change := range changes {
+					log.Printf("Head event { height:%d; type: %s }", change.Val.Height(), change.Type)
+
+					switch change.Type {
+					case store.HCCurrent:
+						tipsets, err := loadTipsets(ctx, api, change.Val, lastHeight)
+						if err != nil {
+							log.Print(err)
+							return
+						}
+
+						for _, tipset := range tipsets {
+							chmain <- tipset
+						}
+					case store.HCApply:
+						chmain <- change.Val
+					}
+				}
+			case <-ping:
+				log.Print("Running health check")
+
+				cctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+				if _, err := api.ID(cctx); err != nil {
+					log.Print("Health check failed")
+					return
+				}
```

Comment:
@whyrusleeping was doing a little bit of sanity checking and it turns out that if the lotus node dies, stats will just hang forever. I originally though this might be due to the fact that everything starts off from this notify channel, so I added this call to `api.ID` (low time durations for speedy reproducibility). However, there doesn't appear to be any timeout handing anywhere, or at least it ignore the context, and I have no idea what the default timeout out. Currently if the lotus node stops / restarts there is no way to detect this and exit / resume on the stats end.

Any ideas?

Reply:
It's a bug in the jsonrpc library we probably should really fix